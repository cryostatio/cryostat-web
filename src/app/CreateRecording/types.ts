/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { EventTemplate, KeyValue } from '@app/Shared/Services/api.types';
import { ValidatedOptions } from '@patternfly/react-core';

export type EventTemplateIdentifier = Pick<EventTemplate, 'name' | 'type'>;

interface _FormBaseData {
  name: string;
  template?: EventTemplateIdentifier;
  labels: KeyValue[];
  continuous: boolean;
  archiveOnStop: boolean;
  restart: boolean;
  duration: number;
  durationUnit: number;
  skipDurationCheck?: boolean;
  maxAge: number;
  maxAgeUnit: number;
  maxSize: number;
  maxSizeUnit: number;
  toDisk: boolean;
}

interface _FormValidationData {
  nameValid: ValidatedOptions;
  labelsValid: ValidatedOptions;
  durationValid: ValidatedOptions;
}

export type CustomRecordingFormData = _FormBaseData & _FormValidationData;
