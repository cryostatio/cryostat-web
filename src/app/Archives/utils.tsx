/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { RecordingDirectory, Target } from '@app/Shared/Services/api.types';

export const includesDirectory = (arr: RecordingDirectory[], dir: RecordingDirectory): boolean => {
  return arr.some((t) => t.connectUrl === dir.connectUrl);
};

export const indexOfDirectory = (arr: RecordingDirectory[], dir: RecordingDirectory): number => {
  let index = -1;
  arr.forEach((d, idx) => {
    if (d.connectUrl === dir.connectUrl) {
      index = idx;
    }
  });
  return index;
};

export const getTargetFromDirectory = (dir: RecordingDirectory): Target => {
  return {
    connectUrl: dir.connectUrl,
    alias: dir.jvmId,
    labels: [],
    annotations: {
      cryostat: [],
      platform: [],
    },
  };
};
