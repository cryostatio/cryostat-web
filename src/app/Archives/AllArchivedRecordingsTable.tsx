/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { ErrorView } from '@app/ErrorView/ErrorView';
import { authFailMessage, isAuthFail } from '@app/ErrorView/types';
import { ArchivedRecordingsTable } from '@app/Recordings/ArchivedRecordingsTable';
import { LoadingView } from '@app/Shared/Components/LoadingView';
import { ArchivedRecording, RecordingDirectory, Target, NotificationCategory } from '@app/Shared/Services/api.types';
import { ServiceContext } from '@app/Shared/Services/Services';
import { useSort } from '@app/utils/hooks/useSort';
import { useSubscriptions } from '@app/utils/hooks/useSubscriptions';
import { TableColumn, portalRoot, sortResources } from '@app/utils/utils';
import {
  Toolbar,
  ToolbarContent,
  ToolbarGroup,
  ToolbarItem,
  SearchInput,
  Badge,
  EmptyState,
  EmptyStateIcon,
  Text,
  Tooltip,
  Split,
  SplitItem,
  EmptyStateHeader,
} from '@patternfly/react-core';
import { HelpIcon, SearchIcon } from '@patternfly/react-icons';
import { Table, Th, Thead, Tbody, Tr, Td, ExpandableRowContent, SortByDirection } from '@patternfly/react-table';
import * as React from 'react';
import { Observable, of } from 'rxjs';
import { getTargetFromDirectory, includesDirectory, indexOfDirectory } from './utils';

const tableColumns: TableColumn[] = [
  {
    title: 'Directory',
    keyPaths: ['connectUrl'],
    sortable: true,
    width: 80,
  },
  {
    title: 'Archives',
    keyPaths: ['recordings'],
    transform: (recordings: ArchivedRecording[], _obj: _RecordingDirectory) => {
      return recordings.length;
    },
    sortable: true,
    width: 15,
  },
];

type _RecordingDirectory = RecordingDirectory & { targetAsObs: Observable<Target> };

export interface AllArchivedRecordingsTableProps {}

export const AllArchivedRecordingsTable: React.FC<AllArchivedRecordingsTableProps> = () => {
  const context = React.useContext(ServiceContext);

  const [directories, setDirectories] = React.useState<_RecordingDirectory[]>([]);
  const [searchText, setSearchText] = React.useState('');
  const [expandedDirectories, setExpandedDirectories] = React.useState<_RecordingDirectory[]>([]);
  const [errorMessage, setErrorMessage] = React.useState('');
  const [isLoading, setIsLoading] = React.useState(false);
  const addSubscription = useSubscriptions();
  const [sortBy, getSortParams] = useSort();

  const handleDirectoriesAndCounts = React.useCallback(
    (directories: RecordingDirectory[]) => {
      setDirectories(directories.map((dir) => ({ ...dir, targetAsObs: of(getTargetFromDirectory(dir)) })));
      setIsLoading(false);
    },
    [setDirectories, setIsLoading],
  );

  const handleError = React.useCallback(
    (error) => {
      setIsLoading(false);
      setErrorMessage(error.message);
    },
    [setIsLoading, setErrorMessage],
  );

  const refreshDirectoriesAndCounts = React.useCallback(() => {
    setIsLoading(true);
    addSubscription(
      context.api.doGet<RecordingDirectory[]>('fs/recordings', 'beta').subscribe({
        next: handleDirectoriesAndCounts,
        error: handleError,
      }),
    );
  }, [addSubscription, context.api, setIsLoading, handleDirectoriesAndCounts, handleError]);

  const handleSearchInput = React.useCallback(
    (_, searchInput: string) => {
      setSearchText(searchInput);
    },
    [setSearchText],
  );

  const handleSearchInputClear = React.useCallback(() => {
    setSearchText('');
  }, [setSearchText]);

  React.useEffect(() => {
    refreshDirectoriesAndCounts();
  }, [refreshDirectoriesAndCounts]);

  const searchedDirectories = React.useMemo(() => {
    let updatedSearchedDirectories: _RecordingDirectory[];
    if (!searchText) {
      updatedSearchedDirectories = directories;
    } else {
      const formattedSearchText = searchText.trim().toLowerCase();
      updatedSearchedDirectories = directories.filter(
        (d: _RecordingDirectory) =>
          d.jvmId.toLowerCase().includes(formattedSearchText) ||
          d.connectUrl.toLowerCase().includes(formattedSearchText),
      );
    }
    return sortResources(
      {
        index: sortBy.index ?? 0,
        direction: sortBy.direction ?? SortByDirection.asc,
      },
      updatedSearchedDirectories,
      tableColumns,
    );
  }, [directories, searchText, sortBy]);

  React.useEffect(() => {
    addSubscription(
      context.target.authFailure().subscribe(() => {
        setErrorMessage(authFailMessage);
      }),
    );
  }, [context, context.target, setErrorMessage, addSubscription]);

  React.useEffect(() => {
    if (!context.settings.autoRefreshEnabled()) {
      return;
    }
    const id = window.setInterval(
      () => refreshDirectoriesAndCounts(),
      context.settings.autoRefreshPeriod() * context.settings.autoRefreshUnits(),
    );
    return () => window.clearInterval(id);
  }, [context.settings, refreshDirectoriesAndCounts]);

  React.useEffect(() => {
    addSubscription(
      context.notificationChannel.messages(NotificationCategory.RecordingMetadataUpdated).subscribe((event) => {
        const updatedRecordingInfo = event.message;

        setDirectories((currentDirectories) => {
          const newDirectories = currentDirectories.map((directory) => ({
            ...directory,
            recordings: directory.recordings.map((recording) => {
              if (recording.name === updatedRecordingInfo.recording.name) {
                return {
                  ...recording,
                  metadata: { ...recording.metadata, labels: updatedRecordingInfo?.recording?.metadata?.labels },
                };
              }
              return recording;
            }),
          }));

          return newDirectories;
        });
      }),
    );
  }, [addSubscription, context.notificationChannel, setDirectories]);

  React.useEffect(() => {
    addSubscription(
      context.notificationChannel.messages(NotificationCategory.ArchivedRecordingCreated).subscribe(() => {
        refreshDirectoriesAndCounts();
      }),
    );
  }, [addSubscription, context.notificationChannel, refreshDirectoriesAndCounts]);

  React.useEffect(() => {
    addSubscription(
      context.notificationChannel.messages(NotificationCategory.ArchivedRecordingDeleted).subscribe(() => {
        refreshDirectoriesAndCounts();
      }),
    );
  }, [addSubscription, context.notificationChannel, refreshDirectoriesAndCounts]);

  const toggleExpanded = React.useCallback(
    (dir) => {
      const idx = indexOfDirectory(expandedDirectories, dir);
      setExpandedDirectories((prevExpandedDirectories) =>
        idx >= 0
          ? [
              ...prevExpandedDirectories.slice(0, idx),
              ...prevExpandedDirectories.slice(idx + 1, prevExpandedDirectories.length),
            ]
          : [...prevExpandedDirectories, dir],
      );
    },
    [expandedDirectories, setExpandedDirectories],
  );

  const directoryRows = React.useMemo(() => {
    return searchedDirectories.map((dir, idx) => {
      const isExpanded: boolean = includesDirectory(expandedDirectories, dir);

      return (
        <Tr key={`${idx}_parent`}>
          <Td
            key={`directory-table-row-${idx}_1`}
            id={`directory-ex-toggle-${idx}`}
            aria-controls={`directory-ex-expand-${idx}`}
            expand={{
              rowIndex: idx,
              isExpanded: isExpanded,
              onToggle: () => toggleExpanded(dir),
            }}
          />
          <Td key={`directory-table-row-${idx}_2`} dataLabel={tableColumns[0].title}>
            <Split hasGutter>
              <SplitItem>
                <Text>{dir.connectUrl}</Text>
              </SplitItem>
              <SplitItem>
                <Tooltip hidden={!dir.jvmId} content={`JVM hash ID: ${dir.jvmId}`} appendTo={portalRoot}>
                  <HelpIcon />
                </Tooltip>
              </SplitItem>
            </Split>
          </Td>
          <Td key={`directory-table-row-${idx}_3`} dataLabel={tableColumns[1].title}>
            <Badge key={`${idx}_count`}>{dir.recordings.length || 0}</Badge>
          </Td>
        </Tr>
      );
    });
  }, [toggleExpanded, searchedDirectories, expandedDirectories]);

  const recordingRows = React.useMemo(() => {
    return searchedDirectories.map((dir, idx) => {
      const isExpanded: boolean = includesDirectory(expandedDirectories, dir);

      return (
        <Tr key={`${idx}_child`} isExpanded={isExpanded}>
          <Td key={`directory-ex-expand-${idx}`} dataLabel={'Content Details'} colSpan={tableColumns.length + 1}>
            {isExpanded ? (
              <ExpandableRowContent>
                <ArchivedRecordingsTable
                  directory={dir}
                  target={dir.targetAsObs}
                  isUploadsTable={false}
                  isNestedTable={true}
                  directoryRecordings={dir.recordings}
                />
              </ExpandableRowContent>
            ) : null}
          </Td>
        </Tr>
      );
    });
  }, [searchedDirectories, expandedDirectories]);

  const rowPairs = React.useMemo(() => {
    const rowPairs: JSX.Element[] = [];
    for (let i = 0; i < directoryRows.length; i++) {
      rowPairs.push(directoryRows[i]);
      rowPairs.push(recordingRows[i]);
    }
    return rowPairs;
  }, [directoryRows, recordingRows]);

  let view: JSX.Element;

  const authRetry = React.useCallback(() => {
    context.target.setAuthRetry();
  }, [context.target]);

  const isError = React.useMemo(() => errorMessage != '', [errorMessage]);

  if (isError) {
    view = (
      <>
        <ErrorView
          title={'Error retrieving Archived Recordings in All Archives View'}
          message={errorMessage}
          retry={isAuthFail(errorMessage) ? authRetry : undefined}
        />
      </>
    );
  } else if (isLoading) {
    view = <LoadingView />;
  } else if (!searchedDirectories.length) {
    view = (
      <>
        <EmptyState>
          <EmptyStateHeader
            titleText="No Archived Recordings"
            icon={<EmptyStateIcon icon={SearchIcon} />}
            headingLevel="h4"
          />
        </EmptyState>
      </>
    );
  } else {
    view = (
      <>
        <Table aria-label="all-archives-table">
          <Thead>
            <Tr>
              <Th key="table-header-expand" />
              {tableColumns.map(({ title, width }, index) => (
                <Th
                  key={`table-header-${title}`}
                  sort={getSortParams(index)}
                  width={width as React.ComponentProps<typeof Th>['width']}
                >
                  {title}
                </Th>
              ))}
            </Tr>
          </Thead>
          <Tbody>{rowPairs}</Tbody>
        </Table>
      </>
    );
  }

  return (
    <>
      <Toolbar id="all-archives-toolbar">
        <ToolbarContent>
          <ToolbarGroup variant="filter-group">
            <ToolbarItem>
              <SearchInput
                placeholder="Search"
                value={searchText}
                onChange={handleSearchInput}
                onClear={handleSearchInputClear}
              />
            </ToolbarItem>
          </ToolbarGroup>
        </ToolbarContent>
      </Toolbar>
      {view}
    </>
  );
};
