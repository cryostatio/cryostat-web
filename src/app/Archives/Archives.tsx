/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { BreadcrumbPage } from '@app/BreadcrumbPage/BreadcrumbPage';
import { ArchivedRecordingsTable } from '@app/Recordings/ArchivedRecordingsTable';
import { Target, UPLOADS_SUBDIRECTORY } from '@app/Shared/Services/api.types';
import { ServiceContext } from '@app/Shared/Services/Services';
import { useSubscriptions } from '@app/utils/hooks/useSubscriptions';
import { getActiveTab, switchTab } from '@app/utils/utils';
import {
  Card,
  CardBody,
  EmptyState,
  EmptyStateIcon,
  Tab,
  Tabs,
  TabTitleText,
  EmptyStateHeader,
} from '@patternfly/react-core';
import { SearchIcon } from '@patternfly/react-icons';
import * as React from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import { of } from 'rxjs';
import { AllArchivedRecordingsTable } from './AllArchivedRecordingsTable';
import { AllTargetsArchivedRecordingsTable } from './AllTargetsArchivedRecordingsTable';
/*
  This specific target is used as the "source" for the Uploads version of the ArchivedRecordingsTable.
  The connectUrl is the 'uploads' because for actions performed on uploaded archived recordings,
  the backend issues a notification with the "target" field set to the 'uploads', signalling that
  these recordings are not associated with any target. We can then match on the 'uploads' when performing
  notification handling in the ArchivedRecordingsTable.
*/
export const uploadAsTarget: Target = {
  connectUrl: UPLOADS_SUBDIRECTORY,
  alias: '',
  labels: [],
  annotations: {
    cryostat: [],
    platform: [],
  },
};

enum ArchiveTab {
  ALL_ARCHIVES = 'all-archives',
  ALL_TARGETS = 'all-targets',
  UPLOADS = 'uploads',
}

export interface ArchivesProps {}

export const Archives: React.FC<ArchivesProps> = ({ ...props }) => {
  const { search, pathname } = useLocation();
  const navigate = useNavigate();
  const context = React.useContext(ServiceContext);
  const addSubscription = useSubscriptions();

  const activeTab = React.useMemo(() => {
    return getActiveTab(search, 'tab', Object.values(ArchiveTab), ArchiveTab.ALL_TARGETS);
  }, [search]);

  const [archiveEnabled, setArchiveEnabled] = React.useState(false);

  React.useEffect(() => {
    addSubscription(context.api.isArchiveEnabled().subscribe(setArchiveEnabled));
  }, [context.api, addSubscription, setArchiveEnabled]);

  const onTabSelect = React.useCallback(
    (_: React.MouseEvent, key: string | number) =>
      switchTab(navigate, pathname, search, { tabKey: 'tab', tabValue: `${key}` }),
    [navigate, pathname, search],
  );

  const uploadTargetAsObs = React.useMemo(() => of(uploadAsTarget), []);

  const cardBody = React.useMemo(() => {
    return archiveEnabled ? (
      <Tabs id="archives" activeKey={activeTab} onSelect={onTabSelect} unmountOnExit>
        <Tab id="all-targets" eventKey={ArchiveTab.ALL_TARGETS} title={<TabTitleText>All Targets</TabTitleText>}>
          <AllTargetsArchivedRecordingsTable />
        </Tab>
        <Tab id="all-archives" eventKey={ArchiveTab.ALL_ARCHIVES} title={<TabTitleText>All Archives</TabTitleText>}>
          <AllArchivedRecordingsTable />
        </Tab>
        <Tab id="uploads" eventKey={ArchiveTab.UPLOADS} title={<TabTitleText>Uploads</TabTitleText>}>
          <ArchivedRecordingsTable target={uploadTargetAsObs} isUploadsTable={true} isNestedTable={false} />
        </Tab>
      </Tabs>
    ) : (
      <EmptyState>
        <EmptyStateHeader
          titleText="Archives Unavailable"
          icon={<EmptyStateIcon icon={SearchIcon} />}
          headingLevel="h4"
        />
      </EmptyState>
    );
  }, [archiveEnabled, activeTab, uploadTargetAsObs, onTabSelect]);

  return (
    <BreadcrumbPage {...props} pageTitle="Archives">
      <Card>
        <CardBody>{cardBody}</CardBody>
      </Card>
    </BreadcrumbPage>
  );
};

export default Archives;
