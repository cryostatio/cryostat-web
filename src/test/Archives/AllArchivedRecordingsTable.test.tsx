/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { AllArchivedRecordingsTable } from '@app/Archives/AllArchivedRecordingsTable';
import { NotificationMessage, ArchivedRecording, RecordingDirectory } from '@app/Shared/Services/api.types';
import { defaultServices } from '@app/Shared/Services/Services';
import '@testing-library/jest-dom';
import { cleanup, screen, within, waitFor } from '@testing-library/react';
import { of } from 'rxjs';
import { render, renderSnapshot } from '../utils';

const mockConnectUrl1 = 'service:jmx:rmi://someUrl1';
const mockJvmId1 = 'fooJvmId1';
const mockConnectUrl2 = 'service:jmx:rmi://someUrl2';
const mockJvmId2 = 'fooJvmId2';
const mockConnectUrl3 = 'service:jmx:rmi://someUrl3';
const mockJvmId3 = 'fooJvmId3';
const mockName3 = 'someRecording3';

const mockCount1 = 1;

const mockRecordingSavedNotification = {
  message: {
    recording: {
      name: mockName3,
      metadata: {
        labels: {
          key: 'someLabel',
          value: 'someValue',
        },
      },
    },
  },
} as NotificationMessage;

const mockRecordingDeletedNotification = {
  message: {
    recording: {
      name: mockName3,
    },
  },
} as NotificationMessage;

const mockRecording: ArchivedRecording = {
  name: 'someRecording',
  downloadUrl: 'http://downloadUrl',
  reportUrl: 'http://reportUrl',
  metadata: {
    labels: [
      {
        key: 'someLabel',
        value: 'someValue',
      },
    ],
  },
  size: 2048,
  archivedTime: 2048,
};

const mockRecordingDirectory1: RecordingDirectory = {
  connectUrl: mockConnectUrl1,
  jvmId: mockJvmId1,
  recordings: [mockRecording],
};

const mockRecordingDirectory2: RecordingDirectory = {
  connectUrl: mockConnectUrl2,
  jvmId: mockJvmId2,
  recordings: [mockRecording],
};

const mockRecordingDirectory3: RecordingDirectory = {
  connectUrl: mockConnectUrl3,
  jvmId: mockJvmId3,
  recordings: [mockRecording, mockRecording, mockRecording],
};

const mockRecordingDirectory3Removed: RecordingDirectory = {
  ...mockRecordingDirectory3,
  recordings: [mockRecording, mockRecording],
};

const mockRecordingDirectory3Added: RecordingDirectory = {
  connectUrl: mockConnectUrl3,
  jvmId: mockJvmId3,
  recordings: [mockRecording, mockRecording, mockRecording, mockRecording],
};

jest.mock('@app/Recordings/ArchivedRecordingsTable', () => {
  return {
    ArchivedRecordingsTable: jest.fn((_) => {
      return <div>Archived Recordings Table</div>;
    }),
  };
});

jest.mock('@app/Shared/Services/Target.service', () => ({
  ...jest.requireActual('@app/Shared/Services/Target.service'), // Require actual implementation of utility functions for Target
}));

jest
  .spyOn(defaultServices.api, 'doGet')
  .mockReturnValueOnce(of([])) // renders correctly

  .mockReturnValueOnce(of([])) // shows no recordings when empty

  .mockReturnValueOnce(of([mockRecordingDirectory1])) // has the correct table elements

  .mockReturnValueOnce(of([mockRecordingDirectory1, mockRecordingDirectory2, mockRecordingDirectory3])) // search function

  .mockReturnValueOnce(of([mockRecordingDirectory1, mockRecordingDirectory2, mockRecordingDirectory3])) // expands targets to show their <ArchivedRecordingsTable />

  // notifications trigger doGet queries
  .mockReturnValueOnce(of([mockRecordingDirectory1, mockRecordingDirectory2, mockRecordingDirectory3])) // increments the count when an archived recording is saved
  .mockReturnValueOnce(of([mockRecordingDirectory1, mockRecordingDirectory2, mockRecordingDirectory3Added]))

  .mockReturnValueOnce(of([mockRecordingDirectory1, mockRecordingDirectory2, mockRecordingDirectory3])) // decrements the count when an archived recording is deleted
  .mockReturnValueOnce(of([mockRecordingDirectory1, mockRecordingDirectory2, mockRecordingDirectory3Removed]));

jest
  .spyOn(defaultServices.notificationChannel, 'messages')
  .mockReturnValueOnce(of()) // renders correctly  // NotificationCategory.RecordingMetadataUpdated
  .mockReturnValueOnce(of()) // NotificationCategory.ArchivedRecordingCreated
  .mockReturnValueOnce(of()) // NotificationCategory.ArchivedRecordingDeleted

  .mockReturnValueOnce(of()) // shows no recordings when empty
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // has the correct table elements
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // correctly handles the search function
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // expands targets to show their <ArchivedRecordingsTable />
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // increments the count when an archived recording is saved
  .mockReturnValueOnce(of(mockRecordingSavedNotification))
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // decrements the count when an archived recording is deleted
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of(mockRecordingDeletedNotification));

describe('<AllArchivedRecordingsTable />', () => {
  afterEach(cleanup);

  it('renders correctly', async () => {
    const tree = await renderSnapshot({
      routerConfigs: { routes: [{ path: '/archives', element: <AllArchivedRecordingsTable /> }] },
    });
    expect(tree?.toJSON()).toMatchSnapshot();
  });

  it('shows no recordings when empty', async () => {
    render({ routerConfigs: { routes: [{ path: '/archives', element: <AllArchivedRecordingsTable /> }] } });

    expect(screen.getByText('No Archived Recordings')).toBeInTheDocument();
  });

  it('has the correct table elements', async () => {
    render({ routerConfigs: { routes: [{ path: '/archives', element: <AllArchivedRecordingsTable /> }] } });

    expect(screen.getByLabelText('all-archives-table')).toBeInTheDocument();
    expect(screen.getByText('Directory')).toBeInTheDocument();
    expect(screen.getByText('Archives')).toBeInTheDocument();
    expect(screen.getByText(`${mockConnectUrl1}`)).toBeInTheDocument();
    expect(screen.getByText('1')).toBeInTheDocument();
  });

  it('correctly handles the search function', async () => {
    const { user } = render({
      routerConfigs: { routes: [{ path: '/archives', element: <AllArchivedRecordingsTable /> }] },
    });

    const search = screen.getByLabelText('Search input');

    let tableBody = screen.getAllByRole('rowgroup')[1];
    let rows = within(tableBody).getAllByRole('row');
    expect(rows).toHaveLength(3);

    await user.type(search, '1');
    tableBody = screen.getAllByRole('rowgroup')[1];
    rows = within(tableBody).getAllByRole('row');
    expect(rows).toHaveLength(1);
    const firstTarget = rows[0];
    expect(within(firstTarget).getByText(`${mockConnectUrl1}`)).toBeTruthy();
    expect(within(firstTarget).getByText(`${mockCount1}`)).toBeTruthy();

    await user.type(search, 'asdasdjhj');
    expect(screen.getByText('No Archived Recordings')).toBeInTheDocument();
    expect(screen.queryByLabelText('all-archives-table')).not.toBeInTheDocument();

    await user.clear(search);
    tableBody = screen.getAllByRole('rowgroup')[1];
    rows = within(tableBody).getAllByRole('row');
    expect(rows).toHaveLength(3);
  });

  it('expands targets to show their <ArchivedRecordingsTable />', async () => {
    const { user } = render({
      routerConfigs: { routes: [{ path: '/archives', element: <AllArchivedRecordingsTable /> }] },
    });

    expect(screen.queryByText('Archived Recordings Table')).not.toBeInTheDocument();

    let tableBody = screen.getAllByRole('rowgroup')[1];
    let rows = within(tableBody).getAllByRole('row');
    expect(rows).toHaveLength(3);

    const firstTarget = rows[0];
    const expand = within(firstTarget).getByLabelText('Details');
    await user.click(expand);

    tableBody = screen.getAllByRole('rowgroup')[1];
    rows = within(tableBody).getAllByRole('row');
    expect(rows).toHaveLength(4);

    const expandedTable = rows[1];
    expect(within(expandedTable).getByText('Archived Recordings Table')).toBeTruthy();

    await user.click(expand);
    tableBody = screen.getAllByRole('rowgroup')[1];
    rows = within(tableBody).getAllByRole('row');
    expect(rows).toHaveLength(3);
    expect(screen.queryByText('Archived Recordings Table')).not.toBeInTheDocument();
  });

  it('increments the count when an archived recording is saved', async () => {
    render({ routerConfigs: { routes: [{ path: '/archives', element: <AllArchivedRecordingsTable /> }] } });

    const tableBody = screen.getAllByRole('rowgroup')[1];
    const rows = within(tableBody).getAllByRole('row');
    expect(rows).toHaveLength(3);

    const thirdTarget = rows[2];
    expect(within(thirdTarget).getByText(`${mockConnectUrl3}`)).toBeTruthy();
    await waitFor(() => {
      expect(within(thirdTarget).getByText('4')).toBeInTheDocument();
    });
  });

  it('decrements the count when an archived recording is deleted', async () => {
    render({ routerConfigs: { routes: [{ path: '/archives', element: <AllArchivedRecordingsTable /> }] } });

    const tableBody = screen.getAllByRole('rowgroup')[1];
    const rows = within(tableBody).getAllByRole('row');

    const thirdTarget = rows[2];
    expect(within(thirdTarget).getByText(`${mockConnectUrl3}`)).toBeTruthy();
    expect(within(thirdTarget).getByText(2)).toBeTruthy();
  });
});
