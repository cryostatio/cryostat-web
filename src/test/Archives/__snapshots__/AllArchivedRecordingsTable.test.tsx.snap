// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AllArchivedRecordingsTable /> renders correctly 1`] = `
Array [
  <div
    className="pf-v5-c-toolbar"
    data-ouia-component-id="OUIA-Generated-Toolbar-1"
    data-ouia-component-type="PF5/Toolbar"
    data-ouia-safe={true}
    id="all-archives-toolbar"
  >
    <div
      className="pf-v5-c-toolbar__content"
    >
      <div
        className="pf-v5-c-toolbar__content-section"
      >
        <div
          className="pf-v5-c-toolbar__group pf-m-filter-group"
        >
          <div
            className="pf-v5-c-toolbar__item"
          >
            <div
              className="pf-v5-c-text-input-group"
            >
              <div
                className="pf-v5-c-text-input-group__main pf-m-icon"
                onKeyDown={[Function]}
              >
                <span
                  className="pf-v5-c-text-input-group__text"
                >
                  <span
                    className="pf-v5-c-text-input-group__icon"
                  >
                    <svg
                      aria-hidden={true}
                      aria-labelledby={null}
                      className="pf-v5-svg"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      viewBox="0 0 512 512"
                      width="1em"
                    >
                      <path
                        d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                      />
                    </svg>
                  </span>
                  <input
                    aria-label="Search input"
                    className="pf-v5-c-text-input-group__text-input"
                    disabled={false}
                    onChange={[Function]}
                    placeholder="Search"
                    type="text"
                    value=""
                  />
                </span>
              </div>
              
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="pf-v5-c-toolbar__content pf-m-hidden"
      hidden={true}
    >
      <div
        className="pf-v5-c-toolbar__group"
      />
    </div>
  </div>,
  <div
    className="pf-v5-c-empty-state"
  >
    <div
      className="pf-v5-c-empty-state__content"
    >
      <div
        className="pf-v5-c-empty-state__header"
      >
        <div
          className="pf-v5-c-empty-state__icon"
        >
          <svg
            aria-hidden={true}
            aria-labelledby={null}
            className="pf-v5-svg"
            fill="currentColor"
            height="1em"
            role="img"
            viewBox="0 0 512 512"
            width="1em"
          >
            <path
              d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            />
          </svg>
        </div>
        <div
          className="pf-v5-c-empty-state__title"
        >
          <h4
            className="pf-v5-c-empty-state__title-text"
          >
            No Archived Recordings
          </h4>
        </div>
      </div>
    </div>
  </div>,
]
`;
